web/modules/custom/joinup_bundle_class/src/JoinupBundleClassMetaEntityTrait.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_bundle_class;
+
+use Drupal\Core\Field\Plugin\Field\FieldType\EntityReferenceItem;
+use Drupal\joinup_bundle_class\Exception\MetaEntityAlreadyExistsException;
+use Drupal\meta_entity\Entity\MetaEntityInterface;
+
+/**
+ * Reusable methods for accessing fields in entity bundle classes.
+ *
+ * @todo This depends on JoinupBundleClassFieldAccessTrait but due to a PHP bug
+ *   this can only be included in PHP 7.3 and higher. Add back the use statement
+ *   once we are on a supported version.
+ * @see https://bugs.php.net/bug.php?id=63911
+ */
+trait JoinupBundleClassMetaEntityTrait {
+
+  /**
+   * Returns the meta entity that is referenced in the given field.
+   *
+   * @param string $field_name
+   *   The name of the computed field that references the meta entity.
+   *
+   * @return \Drupal\meta_entity\Entity\MetaEntityInterface
+   *   The meta entity.
+   *
+   * @throws \InvalidArgumentException
+   *   Thrown when the field does not reference a meta entity.
+   */
+  protected function getMetaEntity(string $field_name): ?MetaEntityInterface {
+    $item = $this->getFirstItem($field_name);
+    if (empty($item)) {
+      return NULL;
+    }
+
+    if (!$item instanceof EntityReferenceItem) {
+      $type = gettype($item) === 'object' ? get_class($item) : gettype($item);
+      $message = sprintf('The field %s on entity %s is expected to be a reference to a meta entity but it is a %s', $field_name, $this->id(), $type);
+      throw new \InvalidArgumentException($message);
+    }
+
+    $target_type = $item->getDataDefinition()->getSetting('target_type') ?? 'unknown';
+    if ($target_type !== 'meta_entity') {
+      $message = sprintf('The field %s on entity %s is expected to reference a meta entity but it references an entity of type %s', $field_name, $this->id(), $target_type);
+      throw new \InvalidArgumentException($message);
+    }
```

Comment:
These `instanceof` checks are mainly intended to catch the `NULL` case, but they are a nice way to reliably process the data from the Field API which can return fields of any type. This serves as a convenient bridge between the Drupal core code (which is not strictly typed) and our own code (which is strictly typed).

The target type check is there to fulfill the promise that we return a `?MetaEntityInterface`. Since we are accepting any field through the `$field_name` parameter it is possible we get a _different_ entity reference in case some dynamic code calls us.

All these checks together ensure that the code returns the expected results, regardless of the input that was passed in.

Regarding your suggestion to use `$entity->pinned_in->entity` - this is not recommended because it can cause `Trying to get property "entity" of non-object`. Disregarding the fact that values can be `NULL` is the source of most bugs we recently encountered in production (see e.g. ISAICP-6102 and ISAICP-6107). This chaining of field API properties is particularly unreliable because any of those elements in the chain can be `NULL`. It is a sad fact that this pattern is so much used across the whole of Drupal and our own code. We should work actively to remove all these instances when we see them.

For more info about why I am adopting this strict practice when dealing with the Field API see the long comment https://github.com/ec-europa/joinup-dev/pull/2190#issuecomment-656281749 of issue ISAICP-6107 where we fixed a fatal error in production which was caused by a very similar line of code:

```
$collection->get('og_audience')->entity
```

Reply:
Well, if `NULL` is an allowed value then it's the caller responsibility to do check:

```php
if ($entity = $entity->pinned_in->entity) {
  // Do something w/ the $entity.
}
```

Reply:
To me all these checks & wrapping methods feel excessive and overdoing. However, I won't block the PR on this.